DM-33360: Improve performance of https as_local() method

airnandez opened the pull request:
## Checklist

- [ ] ran Jenkins

Unit tests executed in my development environment passed. Could not run Jenkins.

timj commented on python/lsst/resources/http.py:
```diff
@@ -313,13 +313,35 @@ def _as_local(self) -> Tuple[str, bool]:
             Always returns `True`. This is always a temporary file.
         """
-        log.debug("Downloading remote resource as local file: %s", self.geturl())
+        log.debug(f"Downloading {self.geturl()} as local file")
```
We are not supposed to use f-strings in any log messages since it is deemed to be more efficient to defer the string interpolation until we know the log message will be issued. I think we can probably delete this log message completely since the `time_this` code below is already going to issue a log message.

airnandez replied:
> My main comment is I don't know why we can't work out the block size for the temp file we did get rather than guess in advance where the temp file will end up.

I think it is useful to have the same size of buffer for both reading from the socket and writing to the temporary file. The size of the temporary file buffer is passed to at the creation of the file (ie. the call to `tempfile.NamedTemporaryFile(...)`), this is why it is computed beforehand. The same size is used to read data from the network.

> I'm intrigued that it's more important to get the local file system block size correct than any kind of network transfer chunk size.

I think both are important to reduce system calls: if I read from the network chunks of 4 KB and write to the temporary file chunks of 1 MB, I will need much more `read` system calls than `write` system calls.

airnandez replied:
> We are not supposed to use f-strings in any log messages...

Understood. I didn't know that.

timj commented on python/lsst/resources/http.py:
```diff
@@ -313,13 +313,35 @@ def _as_local(self) -> Tuple[str, bool]:
+        # temporary directories.
+        buffering = -1
+        chunk_size = None
+        for path in (os.getenv(var) for var in ('TMPDIR', 'TEMP', 'TMP')):
```
Why are we guessing where the file is going to be written? We know where it is going to be written.

airnandez replied:
> Why are we guessing where the file is going to be written? We know where it is going to be written.

These are the locations where `tempfile.mkstemp()` creates the temporary file, if any of them are defined. At this moment in the execution, it is not known which one will be selected, so we have to inspect them in order.

airnandez replied:
This is an excerpt of the documentation of [tempfile.mkstemp](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp)

> _If dir is not None, the file will be created in that directory; otherwise, a default directory is used. The default directory is chosen from a platform-dependent list, but the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables._ 

Should we then enforce, at the application level, that the directory for temporary file creation must be pointed to `TMPDIR` (which is a POSIX-defined variable) the dfault value being `/tmp` ? I find that a cleaner solution than to create every temporary file twice just to retrieve the block size and could even retrieve that block size once and save it.

timj replied:
If we believe that the block size will be fixed for the entire process (modulo shenanigans of tests redefining TMPDIR) then can't we add a class method for doing this where we will create a throwaway temp file once and cache the result?

I suppose another approach is to use the `dir` parameter and define exactly where we want the file to end up -- we can use `tmpdir = os.environ.get("TMPDIR", ResourcePath(".", forceAbsolute=True, forceDirectory=True))` (or do the loop you do now with `cwd` at the end)  and then examine `tmpdir` directly. Still might be worth caching this rather than doing the fsstat call every time.

```python
_TMPDIR = None
def _get_temp_dir():
    if _TMPDIR is None:
        for tmpdir in (...):
            ...
        _TMPDIR = (tmpdir, buffering)
    return _TMPDIR

def _as_local(self):
    tmpdir, buffering = _get_temp_dir()
    ...
```
or something?

timj commented on python/lsst/resources/http.py:
```diff
@@ -313,13 +313,35 @@ def _as_local(self) -> Tuple[str, bool]:
+                break
+
+        with tempfile.NamedTemporaryFile(suffix=self.getExtension(), buffering=buffering,
+                                         delete=False) as tmpFile:
```
Why can't we do the chunk calculation in this block?
```python
fsstats = os.statvfs(tmpFile.name)
```

airnandez replied:
It is possible, but then we could get an asymmetry between the socket buffer size and the temporary file buffer size.

timj replied:
I had missed the fact that the buffering parameter is passed to `NamedTemporaryFile`. 😦

timj commented on python/lsst/resources/http.py:
```diff
@@ -313,13 +313,35 @@ def _as_local(self) -> Tuple[str, bool]:
+
+        with tempfile.NamedTemporaryFile(suffix=self.getExtension(), buffering=buffering,
+                                         delete=False) as tmpFile:
+            log.debug(f"Downloading {int(r.headers['Content-Length'])} bytes from {self.geturl()} "
```
Can you embed this content into the `time_this` call below so we get one informative debug log message for the transfer?

airnandez replied:
I did think of that but thought that the log message with all the details of the chunk size and the remote contents size made sense in debug mode only. If I embed that message in `time_this` will it be systematically shown even in other log levels?

timj replied:
`time_this` can be told what level to use for it's message. It defaults to `logging.DEBUG`.

airnandez replied:
I just checked and confirm that `time_this` only shows the message when the log level is `DEBUG`.

timj replied:
The main difference is that the timer log message has a `timer.` prefix by default -- I doubt that is going to be an issue but if it is an issue maybe you should keep the log output separate.

timj reviewed (changes requested):
You need to run `black` on the file to make it pass the new formatting requirements. I'm a bit confused by the mypy complaint about a missing stub -- I do install the requests type stubs.

My main comment is I don't know why we can't work out the block size for the temp file we did get rather than guess in advance where the temp file will end up.

I'm intrigued that it's more important to get the local file system block size correct than any kind of network transfer chunk size.

timj commented:
Yes, when I relocated the code from daf_butler I lost some test coverage -- specifically there is not a single test of `as_local` because in butler that was being tested explicitly in the datastore. I will need to clean up the tests in this package.

airnandez commented:
> I also forgot to say that I'm concerned that we don't have any of this in the test coverage -- that's probably my fault rather than anything else.

To be clear, in my local development environment I did `scons -j 16` and made sure there were no failing tests. I hope this is what I was expected to do.

timj commented:
Jenkins should work fine with this since the branch is not a fork. The main issue is that we don't actually test `as_local` at all. Maybe you can add a quick test like:

```diff
diff --git a/tests/test_uri.py b/tests/test_uri.py
index ee6eca3..bc016df 100644
--- a/tests/test_uri.py
+++ b/tests/test_uri.py
@@ -938,6 +938,11 @@ class WebdavURITestCase(unittest.TestCase):
         with self.assertRaises(FileNotFoundError):
             self.notExistingFileResourcePath.read()
 
+        with self.existingFileResourcePath.as_local() as local_uri:
+            self.assertTrue(local_uri.isLocal)
+            content = local_uri.read().decode()
+            self.assertEqual(content, "It works!")
+
     @responses.activate
     def testWrite(self):
 
```

airnandez commented:
I implemented the agreed functionality and added the suggested test to `test_uri.py`. I added the possibility of defining the environment variable `LSST_BUTLER_TMPDIR` to specify the location to use for downloading temporary files, in case the user wants that to be different from the standard `TMPDIR`, which is also considered if the former is not defined. If none of them are defined, the current working directory will be used as temporary directory.

I don't quite understand the errors of the `mypy  / lint` check. As far as I can see, the module `requests.packages.urllib3.util.retry` is actually imported by `http.py`

timj commented on python/lsst/resources/http.py:
```diff
@@ -221,6 +221,37 @@ def finalurl(r: requests.Response) -> str:
+    """
+    global _TMPDIR
+    if _TMPDIR:
+        return _TMPDIR
```
It might be worth adding a second `as_local()` test to ensure that we get code coverage on this line.

timj commented on python/lsst/resources/http.py:
```diff
@@ -221,6 +221,37 @@ def finalurl(r: requests.Response) -> str:
+    tmpdir = os.getcwd()
+    for dir in (os.getenv(v) for v in ("LSST_BUTLER_TMPDIR", "TMPDIR")):
+        if dir and os.path.isdir(dir):
+            tmpdir = dir
```
The implication from the code coverage is that neither of these environment variables are set. What you can do is in the test do something like:

```python
with unittest.mock.patch.dict(os.environ, {"TMPDIR": TESTDIR}):
    with self.existingFileResourcePath.as_local() as local_uri:
        ...
        self.assertIsNotNone(local_uri.relative_to(ResourcePath(TESTDIR)))
```
which would check that the local URI is a subdirectory of TESTDIR.

timj replied:
It occurs to me that this is not going to be reliable because we cache the `_TEMPDIR` and so by definition are not allowing the answer to change between calls. I guess the test could force the `_TEMPDIR` global to be `None` before re-running the test (it's allowed to know about the implementation).

timj commented on python/lsst/resources/http.py:
```diff
@@ -221,6 +221,37 @@ def finalurl(r: requests.Response) -> str:
+    # Use the value of environment variables 'LSST_BUTLER_TMPDIR' or
+    # 'TMPDIR', if defined. Otherwise use current working directory
+    tmpdir = os.getcwd()
+    for dir in (os.getenv(v) for v in ("LSST_BUTLER_TMPDIR", "TMPDIR")):
```
This package is nothing to do with Butler so it's not an obvious name. I know that `http.py` has plenty of other environment variables that do but I didn't change them because I was worried I would break your environment. I think this should probably be `LSST_RESOURCES_TMPDIR` (I do use a `LSST_RESOURCES_TEST_TMP` in tests but you wouldn't use that).

timj reviewed (approved):
I'm happy with this. My main comments concern test coverage.

timj commented:
@airnandez I fixed the mypy problem and pushed a fix.

timj commented:
@airnandez I also added a couple of extra tests so we get full coverage of the patch. I'm happy with this now. I'll probably have to fix the S3 code path now to make it match the `LSST_RESOURCES_TMPDIR` usage. Please take a look and I can merge this if you are happy.

airnandez commented:
> I know that http.py has plenty of other environment variables that do but I didn't change them because I was worried I would break your environment.

That is the reason I choose the `LSST_BUTLER_...` prefix, but you are right that the `resources` module is not specific to butler, so that name is not appropriate.

Changing the other `LSST_BUTLER_*`  environment variables would break things for us but I think it is better to break them now and adapt than to live with that forever. I propose we address those changes in a separate pull request I would be happy to contribute, if you agree.

airnandez commented:
Thanks a lot for all these final touches (and for your patience!). I am happy with you merging these changes.

timj commented:
> I propose we address those changes in a separate pull request I would be happy to contribute, if you agree.

Works for me.

timj commented:
I forgot to answer this question:

> To be clear, in my local development environment I did scons -j 16 and made sure there were no failing tests. I hope this is what I was expected to do.

Yes, but my comment referred to the coverage report on this pull request where it showed which lines were not covered by tests.
